--- src/node.cc	2022-08-26 13:05:24.458396341 -0500
+++ src/node.cc	2022-08-26 13:06:12.179824998 -0500
@@ -479,6 +479,13 @@
     return scope.EscapeMaybe(cb(info));
   }
 
+  // To allow people to extend Node in different ways, this hook allows
+  // one to drop a file lib/_third_party_main.js into the build
+  // directory which will be executed instead of Node's normal loading.
+  if (BuiltinLoaderEnv::Exists("_third_party_main")) {
+    return StartExecution(env, "internal/main/run_third_party_main");
+  }
+
   if (env->worker_context() != nullptr) {
     return StartExecution(env, "internal/main/worker_thread");
   }
