diff --git a/src/heap/cppgc/page-memory.cc b/src/heap/cppgc/page-memory.cc
index 49019ef..057f236 100644
--- a/deps/v8/src/heap/cppgc/page-memory.cc
+++ b/deps/v8/src/heap/cppgc/page-memory.cc
@@ -194,13 +194,9 @@
   auto* pmr = page_memory_region_tree_.Lookup(writeable_base);
   DCHECK_NOT_NULL(pmr);
   page_memory_region_tree_.Remove(pmr);
-  page_pool_.Add(pmr);
-  if (free_memory_handling == FreeMemoryHandling::kDiscardWherePossible) {
-    CHECK(TryDiscard(normal_page_allocator_, pmr->GetPageMemory()));
+  if (free_memory_handling == FreeMemoryHandling::kDoNotDiscard) {
+    page_pool_.Add(pmr);
+  } else {
+    // Destroy the page right away.
+    auto size = normal_page_memory_regions_.erase(pmr);
+    USE(size);
+    DCHECK_EQ(1u, size);
   }
 }
 
