--- a/deps/v8/include/v8config.h
+++ b/deps/v8/include/v8config.h
@@ -491,8 +491,9 @@
 # define V8_ASSUME USE
 #endif
 
-// Prefer c++20 std::assume_aligned
-#if __cplusplus >= 202002L && defined(__cpp_lib_assume_aligned)
+// Prefer c++20 std::assume_aligned. Don't use it on MSVC though, because it's
+// not happy with our large 4GB alignment values.
+#if __cplusplus >= 202002L && defined(__cpp_lib_assume_aligned) && !V8_CC_MSVC
 # define V8_ASSUME_ALIGNED(ptr, alignment) \
   std::assume_aligned<(alignment)>(ptr)
 #elif V8_HAS_BUILTIN_ASSUME_ALIGNED
