--- /dev/null	2022-08-18 11:11:24.665352687 -0500
+++ lib/internal/main/run_third_party_main.js	2022-08-26 13:55:53.482283827 -0500
@@ -0,0 +1,13 @@
+'use strict';
+
+const {
+  prepareMainThreadExecution
+} = require('internal/bootstrap/pre_execution');
+
+prepareMainThreadExecution();
+markBootstrapComplete();
+
+// Legacy _third_party_main.js support
+process.nextTick(() => {
+  require('_third_party_main');
+});
