--- a/src/node.cc
+++ b/src/node.cc
@@ -407,6 +407,13 @@ MaybeLocal<Value> StartExecution(Environment* env, StartExecutionCallback cb) {
   if (env->worker_context() != nullptr) {
     return StartExecution(env, "internal/main/worker_thread");
   }
+
+  // To allow people to extend Node in different ways, this hook allows
+  // one to drop a file lib/_third_party_main.js into the build
+  // directory which will be executed instead of Node's normal loading.
+  if (env->builtin_loader()->Exists("_third_party_main")) {
+    return StartExecution(env, "internal/main/run_third_party_main");
+  }

   std::string first_argv;
   if (env->argv().size() > 1) {
